c/blake3.h

```diff
@@ -38,20 +60,20 @@ typedef struct {
   uint8_t cv_stack[(BLAKE3_MAX_DEPTH + 1) * BLAKE3_OUT_LEN];
 } blake3_hasher;
 
-const char *blake3_version(void);
-void blake3_hasher_init(blake3_hasher *self);
-void blake3_hasher_init_keyed(blake3_hasher *self,
-                              const uint8_t key[BLAKE3_KEY_LEN]);
-void blake3_hasher_init_derive_key(blake3_hasher *self, const char *context);
-void blake3_hasher_init_derive_key_raw(blake3_hasher *self, const void *context,
-                                       size_t context_len);
-void blake3_hasher_update(blake3_hasher *self, const void *input,
-                          size_t input_len);
-void blake3_hasher_finalize(const blake3_hasher *self, uint8_t *out,
-                            size_t out_len);
-void blake3_hasher_finalize_seek(const blake3_hasher *self, uint64_t seek,
-                                 uint8_t *out, size_t out_len);
-void blake3_hasher_reset(blake3_hasher *self);
+BLAKE3_API const char *blake3_version(void);
+BLAKE3_API void blake3_hasher_init(blake3_hasher *self);
+BLAKE3_API void blake3_hasher_init_keyed(blake3_hasher *self,
+                                         const uint8_t key[BLAKE3_KEY_LEN]);
+BLAKE3_API void blake3_hasher_init_derive_key(blake3_hasher *self, const char *context);
+BLAKE3_API void blake3_hasher_init_derive_key_raw(blake3_hasher *self, const void *context,
+                                                  size_t context_len);
+BLAKE3_API void blake3_hasher_update(blake3_hasher *self, const void *input,
+                                     size_t input_len);
+BLAKE3_API void blake3_hasher_finalize(const blake3_hasher *self, uint8_t *out,
+                                       size_t out_len);
+BLAKE3_API void blake3_hasher_finalize_seek(const blake3_hasher *self, uint64_t seek,
+                                            uint8_t *out, size_t out_len);
+BLAKE3_API void blake3_hasher_reset(blake3_hasher *self);
```

Comment:
Could you say more about what these macros are doing? I'm a little nervous about leaking complexity into the header files that might affect folks who aren't using CMake to build this.

Reply:
These macros control which symbols (=> C functions / global variables) end up in the shared object's (.so) symbol table, i.e. what you can access via [`dlsym()`](https://man7.org/linux/man-pages/man3/dlsym.3.html). The default on gcc and clang is to export all entities with external linkage while the Windows / MSVC defaults to exporting only entities annotated with `__declspec(dllexport)`. Without these annotations you basically can't use ther blake3.dll. So these aren't actually CMake specific. They are useful regardless of the buildsystem. You can see a similar macro used by libb2 here:
https://github.com/BLAKE2/libb2/blame/6fafe5c546860ecbdd2b97093a93ae4ce54dfe52/src/blake2.h#L20-L44

You can read more about symbol visibility on the [GCC Wiki](https://gcc.gnu.org/wiki/Visibility) and the [MSVC docs](https://learn.microsoft.com/en-us/cpp/build/exporting-from-a-dll-using-declspec-dllexport?view=msvc-170).

Also note that I configured CMake to set the default visibility to hidden for GCC / clang.